sdk/src/main/scala/com/horizen/AbstractSidechainNodeViewHolder.scala

```diff
@@ -245,22 +247,100 @@ abstract class AbstractSidechainNodeViewHolder[
                                     stateToApply: MS,
                                     walletToApply: VL,
                                     suffixTrimmed: IndexedSeq[PMOD],
-                                    progressInfo: ProgressInfo[PMOD]): Try[SidechainNodeUpdateInformation]
+                                    progressInfo: ProgressInfo[PMOD]): Try[SidechainNodeUpdateInformation] =  Try {
+
+    val updateInfoSample = SidechainNodeUpdateInformation(history, stateToApply, walletToApply, None, None, suffixTrimmed)
+
+    val wrappedResultObj : Try[SidechainNodeUpdateInformation] = progressInfo.toApply.foldLeft[Try[SidechainNodeUpdateInformation]](Success(updateInfoSample)) {
+      case (f@Failure(ex), _) =>
+        log.error("Reporting modifier failed", ex)
+        f
+      case (success@Success(updateInfo), modToApply) =>
+        if (updateInfo.failedMod.isEmpty) {
+          // Check if the next modifier will change Consensus Epoch, so notify History and Wallet with current info.
+          val (newHistory, newWallet) = if (updateInfo.state.isSwitchingConsensusEpoch(modToApply.timestamp)) {
+            log.debug("Switching consensus epoch")
+            val (lastBlockInEpoch, consensusEpochInfo) = updateInfo.state.getCurrentConsensusEpochInfo
+            val nonceConsensusEpochInfo = updateInfo.history.calculateNonceForEpoch(blockIdToEpochId(lastBlockInEpoch))
```

Comment:
I think we still not processing properly the exception here.
We should also notify history similar to the code below:
```
case Failure(e) => {
              log.error(s"Could not apply modifier ${modToApply.id}, exception:" + e)
              newHistory.reportModifierIsInvalid(modToApply, progressInfo).map {
                case (historyAfterApply, newProgressInfo) =>
                  context.system.eventStream.publish(SemanticallyFailedModification(modToApply, e))
                  SidechainNodeUpdateInformation(historyAfterApply, updateInfo.state, newWallet, Some(modToApply), Some(newProgressInfo), updateInfo.suffix)
              }
```

So, now we throw an exception that is wrapped, but:
1) we don't notify history that this block is invalid
2) moreover we don't notify NVH with the proper `SidechainNodeUpdateInformation` instance, so don't revert to state to the proper tip

Reply:
Done